text/3680-use.md

```diff
@@ -0,0 +1,463 @@
+of a `Copy` type ignore the `Clone` implementation and just copy?)
+
+We could make `x.use` inside of a closure cause the closure to get a new use of
+`x` even if it otherwise could borrow, effectively "floating" the `use` outside
+of the closure. This would have different semantics, such as if the closure
+mutated the object before the `use`, so deciding this is a one-way door.
+
+# Prior art
+[prior-art]: #prior-art
+
+Many languages have built-in reference counting, and automatically manage
+reference counts when copying or passing around objects. `.use` provides a
+simple and efficient way for Rust to manage reference counts.
+
+Some languages have copy constructors, allowing arbitrary code to run when
+copying an object. Such languages can manage reference-counted smart pointers
+implicitly.
+
+Rust already has other uses of postfix syntax, notably `.await` and `?`. In
+particular, `.await` provides precedent for the use of `.keyword`.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Should we allow *any* `x.clone()` call to be elided if a type is `Use` (or `Use
++ Copy`)? If we did this, that would not give us a way to explicitly avoid
+elision (unless we added a separate mechanism for that), but it would also mean
+that we could avoid tracking whether a clone call came from `use` or not when
+applying the elision rules, which might make the implementation simpler. (This
+RFC proposes not allowing explicit `x.clone()` calls to be elided, so that
+users can make explicit `x.clone()` calls to avoid elision if desired.)
+
+Are there use cases for using something like `ToOwned` rather than always going
+from `&self` to `Self`? Would any smart pointers need that?
+
+Should `Use` be a `#[marker]` trait?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+We could implement `Use` for small arrays (e.g. arrays smaller than 64 bytes).
+
+We could consider using lints or similar to help avoid unexpected expensive
+copies of *large* arrays or other large `Copy` types, potentially using `Use`
+as a marker for what is not expensive. More generally, we could explore the
+space of types with various combinations of `Use` and `Copy`, and what we want
+`Copy + Use` vs `Copy + !Use` to indicate.
+
+We could allow `.use` in struct construction without duplicating the name of a
+field. For instance:
+
+```rust
+let s = SomeStruct { field, field2.use, field3.use };
+/// This expands to:
+let s = SomeStruct {
+    field: field,
+    field2: field2.use,
+    field3: field3.use,
+};
+```
+
+We could extend the `use` syntax on closures and async blocks to support
+naming specific objects to use:
+
+```rust
+use(x) || { ... }
+
+async use(x) { ... }
+```
+
+We could further extend the `use` syntax to support an explicit capture list of
+named expressions:
+
+```rust
+use(x = x.method(), y) || { ... }
+
+// `..` additionally captures everything that a plain `use` would
+async use(x = &obj, move y, ..)
```

Comment:
I really like this idea. If someone in the project wanted complete transparency, an option could be added to Clippy to prevent the use of use || { ... } without explicitly specifying the variables that should be cloned (or moved).